### PR TITLE
feat: JsonSchemaRegistrySerializer, ProtobufSchemaRegistrySerializer

### PR DESCRIPTION
**Pull request by fab60**
# Description

A Protobuf and Json message serializer fully integrated with the confluent schema registry. Dependencies have been updated to 1.6.3 for several confluent packages

Fixes # (issue) [138](https://github.com/Farfetch/kafka-flow/issues/138)

## How Has This Been Tested?
Integration tests provided for both components

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [ ] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

File: src/KafkaFlow.Serializer.ProtoBuf/KafkaFlow.Serializer.ProtoBuf.csproj [hunk 1]
```diff
@@ -9,10 +9,14 @@
...
   <ItemGroup>
+    <PackageReference Include="Confluent.Kafka" Version="1.6.3" />
+    <PackageReference Include="Confluent.SchemaRegistry" Version="1.6.3" />
+    <PackageReference Include="Confluent.SchemaRegistry.Serdes.Protobuf" Version="1.6.3" />
     <PackageReference Include="protobuf-net" Version="2.4.6" />
   </ItemGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\KafkaFlow.Abstractions\KafkaFlow.Abstractions.csproj" />
+    <ProjectReference Include="..\KafkaFlow.Serializer\KafkaFlow.Serializer.csproj" />
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer.ProtoBuf/KafkaFlow.Serializer.ProtoBuf.csproj:20]: Please, undo the changes in this file

**Review by filipeesch — CHANGES_REQUESTED** [review]
First of all, thanks for your collaboration 😃 

There are some changes to be made here
-  Create two new packages: KafkaFlow.Serializers.ConfluentJsonSerializer and KafkaFlow.Serializers.ConfluentProtobufSerializer (look our Avro csproj as a template)
- Put the new serializers inside these packages
- Create another git commit only with the packages upgrades

File: src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace KafkaFlow.Serializer.Json
+{
+    /// <summary>
+    /// A json message serializer integrated with the confluent schema registry
+    /// </summary>
+    public class JsonSchemaRegistrySerializer : IMessageSerializer
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs:14]: On another hand, maybe it could be nice to have explicit 'SchemaRegistry' in the name of the serializers that are integrated with it.

Another suggestion could be ConfluentJsonSchemaRegistrySerializer.

File: src/KafkaFlow.Serializer.ProtoBuf/ProtobufSchemaRegistrySerializer.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace KafkaFlow.Serializer.ProtoBuf
+{
+    /// <summary>
+    /// A protobuf message serializer integrated with the confluent schema registry
+    /// </summary>
+    public class ProtobufSchemaRegistrySerializer : IMessageSerializer
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.ProtoBuf/ProtobufSchemaRegistrySerializer.cs:13]: Like said [here](https://github.com/Farfetch/kafka-flow/pull/140/files#r615702315), another suggestion could be ConfluentProtobufSchemaRegistrySerializer.

File: src/KafkaFlow.Serializer/KafkaFlow.Serializer.csproj [hunk 4]
```diff
@@ -7,6 +7,10 @@
     <PackageIconUrl>https://raw.githubusercontent.com/Farfetch/.github/master/images/fuse-logo-128.png</PackageIconUrl>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Confluent.SchemaRegistry" Version="1.6.3" />
+  </ItemGroup>
+
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer/KafkaFlow.Serializer.csproj:13]: We are trying to reference external libraries only when is strictly necessary.
If we add 'Confluent.SchemaRegistry' library in this project, when someone wants to use only the 'NewtonsoftJsonMessageSerializer', for example, the 'Confluent.SchemaRegistry' would be added to the project as well (and wouldn´t be used).

File: src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs [hunk 5]
```diff
@@ -6,15 +6,16 @@
     using Confluent.Kafka.SyncOverAsync;
     using Confluent.SchemaRegistry;
     using Confluent.SchemaRegistry.Serdes;
-    
+    using KafkaFlow.Serializer;
```
Comment (dougolima) [inline src/KafkaFlow.Serializer.ApacheAvro/ApacheAvroMessageSerializer.cs:9]: I think this is not necessary , can you check please?

File: src/KafkaFlow.Serializer.Json/KafkaFlow.Serializer.Json.csproj [hunk 6]
```diff
@@ -9,10 +9,13 @@
...
 
   <ItemGroup>
+    <PackageReference Include="Confluent.SchemaRegistry" Version="1.6.3" />
+    <PackageReference Include="Confluent.SchemaRegistry.Serdes.Json" Version="1.6.3" />
     <PackageReference Include="System.Text.Json" Version="4.7.2" />
   </ItemGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\KafkaFlow.Abstractions\KafkaFlow.Abstractions.csproj" />
+    <ProjectReference Include="..\KafkaFlow.Serializer\KafkaFlow.Serializer.csproj" />
```
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.Json/KafkaFlow.Serializer.Json.csproj:19]: Please see [this](https://github.com/Farfetch/kafka-flow/pull/140/files#r615706263).

File: src/KafkaFlow.Serializer.ProtoBuf/KafkaFlow.Serializer.ProtoBuf.csproj (hunk 1 above)
Comment (dougolima) [reply src/KafkaFlow.Serializer.ProtoBuf/KafkaFlow.Serializer.ProtoBuf.csproj:20]: Please see [this](https://github.com/Farfetch/kafka-flow/pull/140/files#r615706263).

File: src/KafkaFlow.Serializer/ConfigurationExtensions.cs [hunk 7]
```diff
@@ -1,10 +1,9 @@
-namespace KafkaFlow.Serializer.ApacheAvro
+namespace KafkaFlow.Serializer
```
Comment (dougolima) [inline src/KafkaFlow.Serializer/ConfigurationExtensions.cs:1]: Please see [this comment](https://github.com/Farfetch/kafka-flow/pull/140/files#r615706263).

**Review by dougolima — CHANGES_REQUESTED** [review]

File: src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (fab60) [reply src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs:14]: I would second @dougolima suggestion to have an explicit reference to 'SchemaRegistry': ConfluentJsonSchemaRegistrySerializer

File: src/KafkaFlow.Serializer/KafkaFlow.Serializer.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (fab60) [reply src/KafkaFlow.Serializer/KafkaFlow.Serializer.csproj:13]: That reference was introduced in order to have shared implementation of the **ClusterConfigurationBuilderExtensions** that was previously defined only on the AvroSerializer project

File: src/KafkaFlow.Serializer.Json/KafkaFlow.Serializer.Json.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (fab60) [reply src/KafkaFlow.Serializer.Json/KafkaFlow.Serializer.Json.csproj:19]: Same reason as before: one possible solution to the problem is to have a single project for all confluent based serializer

File: src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs:14]: I think the name ConfluentJsonSchemaRegistrySerializer is too big and as we have no option to use or not the Schema Registry in these serializers, I think there is no reason to put the name in there. Just my opinion 😄 

I think the users should know the characteristics of each serializer, we must have some documentation with it. Initially, I thought in something like this

| KafkaFlow Serializer                                                                                                | Serialization                                                                                   | Client Package					          | Scema Registry
| ------------------------------------------------------------------------------------------------------------------ | ----------------------------------------------------------------------------------------------- | ---------------------------------------- | -------------------- |
| [KafkaFlow.Serializer.Json](https://www.nuget.org/packages/KafkaFlow.Serializer.Json/)                             | JSON                                                                                            | System.Text.Json                         | No
| [KafkaFlow.Serializer.ConfluentJson](https://www.nuget.org/packages/KafkaFlow.Serializer.Json/)                    | JSON                                                                                            | Confluent.SchemaRegistry.Serdes.Json     | Yes

File: src/KafkaFlow.Serializer/KafkaFlow.Serializer.csproj (hunk 4 above)
[the remark this answers is not in the document]
Comment (filipeesch) [reply src/KafkaFlow.Serializer/KafkaFlow.Serializer.csproj:13]: @fab60 I think you can create a KafkaFlow.SchemaRegistry project e put this class there.

File: src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.Json/JsonSchemaRegistrySerializer.cs:14]: Yeah, the name is big. However, the benefit is that only the reading the name of the serializer you will know it does.

I agree the clients should know the characteristics of the serializers and we need to have a wiki page about it as well, but I think this doesn't avoid us having a name that would ease the understanding.

IMHO, the pattern could be [**PACKAGE-USED**(Confluent, Core,etc) ][**TYPE OR SERIALIZATION**(json, avro, etc)][**SCHEMAREGISTRY**(or not)]**Serializer**.

File: src/KafkaFlow.Serializer.Json/KafkaFlow.Serializer.Json.csproj (hunk 6 above)
[the remark this answers is not in the document]
Comment (dougolima) [reply src/KafkaFlow.Serializer.Json/KafkaFlow.Serializer.Json.csproj:19]: Yes, I think we can have a dedicated Kafkaflow.Serialization.Confluent project to be responsible for registering the confluent based serializers.

Basically, it would have only the 'WithSchemaRegistry' method, at least for now.

**Comment by fab60** [comment]
It seems that integration tests are failing due to the plaintext security protocol being currently configured. Not sure if automatic topic creation is also available in the current cluster configuration.

File: src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard2.0</TargetFramework>
```
Comment (filipeesch) [inline src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj:4]: Missing PackageId, Description and PackageIconUrl tags

File: src/KafkaFlow.Serializer.ProtoBuf/KafkaFlow.Serializer.ProtoBuf.csproj [hunk 9]
```diff
@@ -9,6 +9,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="Confluent.Kafka" Version="1.6.3" />
```
Comment (filipeesch) [inline src/KafkaFlow.Serializer.ProtoBuf/KafkaFlow.Serializer.ProtoBuf.csproj:12]: I think this file should not be changed 😃

File: src/KafkaFlow.Admin.WebApi/Properties/launchSettings.json [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    "KafkaFlow.Admin.WebApi": {
+      "commandName": "Project",
+      "launchBrowser": true,
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      },
+      "applicationUrl": "https://localhost:5001;http://localhost:5000"
+    }
+  }
+}
```
Comment (filipeesch) [inline src/KafkaFlow.Admin.WebApi/Properties/launchSettings.json:27]: Please, remove this file 😃

**Review by filipeesch — CHANGES_REQUESTED** [review]

**Comment by filipeesch** [comment]
@fab60 the topics for the integration tests should be created in the file https://github.com/Farfetch/kafka-flow/blob/master/Makefile

File: src/KafkaFlow.IntegrationTests/Core/Messages/TestMessage3.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+﻿namespace KafkaFlow.IntegrationTests.Core.Messages
+{
+    using System;
+
+    public class TestMessage3 : ITestMessage
+    {
+        public TestMessage3() { }
```
Comment (filipeesch) [inline src/KafkaFlow.IntegrationTests/Core/Messages/TestMessage3.cs:7]: You can remove this ctor

**Review by filipeesch — CHANGES_REQUESTED** [review]

File: src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj (hunk 8 above)
Comment (fab60) [reply src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj:4]: Is this project intended to be packaged as a nuget? I think it does only provide internal services

Comment (filipeesch) [reply src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj:4]: Yes, it is.
If we wouldn't publish it on nuget it won't be available to the client applications and the other packages that use it won't be able to be installed 😃

Comment (fab60) [reply src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj:4]: I think the assembly will be simply packaged as a direct dependency inside the serializer’s package. It would be rather confusing if it will be published autonomously

Comment (filipeesch) [reply src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj:4]: As this package will be a dependency of the other three packages, it is better to publish it autonomously.  Because we can have a package conflict between different serialization types. Imagine if an application uses the ConfluentProtobuf and ConfluentAvro with different package versions, which version of the SchemaRegistry will be used? We need to let the client choose what version of the package they want to use.

Another situation is where the client application wants to create your own serialization with schema registry support, they can use the SchemaRegistry package only.

File: src/KafkaFlow.SchemaRegistry/ClusterConfigurationBuilderExtensions.cs [hunk 12]
```diff
@@ -1,10 +1,12 @@
-namespace KafkaFlow.Serializer.ApacheAvro
+namespace KafkaFlow.SchemaRegistry
```
Comment (filipeesch) [inline src/KafkaFlow.SchemaRegistry/ClusterConfigurationBuilderExtensions.cs:1]: You can use here the namespace `KafkaFlow` only and set the root namespace of the project as `KafkaFlow` only too

**Review by filipeesch — CHANGES_REQUESTED** [review]

File: src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj (hunk 8 above)
Comment (fab60) [reply src/KafkaFlow.SchemaRegistry/KafkaFlow.SchemaRegistry.csproj:4]: I understand your position, but consider that the project is currently exposing nothing more than an extension method on the **IClusterConfigurationBuilder**. There is no public API effectivily exposed that's in some form related to the SchemaRegistry.

As you are the maintener of the project I will definetly follow your guidance.

**Comment by dougolima** [comment]
Hi @fab60 thanks again for your valuable contribution. 👏 👏 👏 

We have just approved your PR and this is going live on the next major version of Kafkaflow.

We are already working on v2 but it can take some time still.

If you need these features immediately, please let me know and we will create a beta version of v2 for you.
